rfc/001_constraintsV2.md

```diff
@@ -0,0 +1,172 @@
+on it.
+
+
+## Changes
+
+Current (V1) structure is:
+
+```yaml
+packages:
+  <maintainer>
+    - package1 (optional constraint)
+  <other section>
+    - package1 (optional constraint)
+expected-test-failures:
+  - package1
+etc
+etc
+etc
+```
+
+I propose that we change this to centralize information about a package (V2), e.g. for me this could look like this:
+
+```yaml
+packages:
+  "Adam Bergmark <adam@bergmark.nl> @bergmark":
+    - name: aeson
+      tests: skip # QuickCheck-2.11.3, base-orphans-0.7, hashable-time
+    - name: HUnit
+      range: < 1.6.2.0 # https://github.com/commercialhaskell/stackage/issues/5834
+    - attoparsec-iso8601
+    - name: fay
+      range: < 0
+      flags:
+        test: true
+    - name: fay-base
+      hide: true
+      range: < 0
+    - name: fay-dom
+      range: < 0
+    - feed
+    - time-compat
+      - tests: skip # base-compat 0.11
+    - through-text
+    # Not my packages
+    - HStringTemplate
+    - name: language-ecmascript
+      range: < 0 # via Diff-0.4.0
+      tests: skip # testing-feat 1.1.0.0
+    - spoon
+    - tagshare
+```
+
+## Downsides
+
+From a maintainer perspective I only see upsides for this.
+
+For curators it will be more work when we need to upper bound lots of
+packages as we can't add everything in one place. But some of us seem
+to already prefer this flow as we are using a mix of the "Stackage
+upper bounds" section and inlining this under the maintainers. Notably
+GHC upgrades would be more time consuming as we always have a large
+chunk of packages that need to be disabled. We could still allow this,
+but I would prefer if this was seen as a temporary measure that we
+clean up after the upgrade.
+
+You won't as easily be able to sort the packages by name. Hopefully
+I'm the only one who feels the urge to sort them.
+
+## Implementation
+
+This is actually very straight forward as curator's internal
+representation maps closely to this new format (the old format
+requires jumping through a lot more hoops to do the conversion).  If
+we decide to support a mix of V1 and V2 style then it will of course
+be more complex, but not that hard since we already have the V1
+implementation to steal from.
+
+## Migration
+
+1. We can implement a function that converts V1 to V2 to transform all
+of build-constraints.yaml.
+1. We could do this (semi?) manually by supporting the V1 & V2 mix and
+moving things a little bit at a time.
+1. A mix of the two above.
+
+## Additional Features
+
+
+### Supporting mass operations
+We could still support the old way of having a group of upper bounds
+in one place, to speed up the work needed on GHC upgrades and package
+relaeses with big impact. This will make the format more complex than
+what V1 is, but apart from the initial work needed I don't think
+that's much of an issue. I would prefer to not have to do this... but
+I think it's necessary to keep people sane.
+
+EDIT: Another idea which I would prefer. Extend `curator` with some
```

Comment:
@danburton @mihaimaruseac since you have been doing a lot of the GHC upgrading chores, do you think this would be good?

Reply:
+2 to helper commands. It's much easier than scrolling through a large file